contrib/TestHarness2/test_harness/run.py

```diff
@@ -0,0 +1,469 @@
+from __future__ import annotations
+
+import array
+import base64
+import collections
+import math
+import os
+import resource
+import shutil
+import subprocess
+import re
+import sys
+import threading
+import time
+import uuid
+
+from functools import total_ordering
+from pathlib import Path
+from test_harness.version import Version
+from test_harness.config import config
+from typing import List, Pattern, OrderedDict
+
+from test_harness.summarize import Summary, SummaryTree
+
+
+@total_ordering
+class TestDescription:
+    def __init__(self, path: Path, name: str, priority: float):
+        self.paths: List[Path] = [path]
+        self.name = name
+        self.priority: float = priority
+        # we only measure in seconds. Otherwise, keeping determinism will be difficult
+        self.total_runtime: int = 0
+        self.num_runs: int = 0
+
+    def __lt__(self, other):
+        if isinstance(other, TestDescription):
+            return self.name < other.name
+        else:
+            return self.name < str(other)
+
+    def __eq__(self, other):
+        if isinstance(other, TestDescription):
+            return self.name < other.name
+        else:
+            return self.name < str(other.name)
+
+
+class StatFetcher:
+    def __init__(self, tests: OrderedDict[str, TestDescription]):
+        self.tests = tests
+
+    def read_stats(self):
+        pass
+
+    def add_run_time(self, test_name: str, runtime: int, out: SummaryTree):
+        self.tests[test_name].total_runtime += runtime
+
+
+class TestPicker:
+    def __init__(self, test_dir: Path):
+        if not test_dir.exists():
+            raise RuntimeError('{} is neither a directory nor a file'.format(test_dir))
+        self.include_files_regex = re.compile(config.include_test_files)
+        self.exclude_files_regex = re.compile(config.exclude_test_files)
+        self.include_tests_regex = re.compile(config.include_test_names)
+        self.exclude_tests_regex = re.compile(config.exclude_test_names)
+        self.test_dir: Path = test_dir
+        self.tests: OrderedDict[str, TestDescription] = collections.OrderedDict()
+        self.restart_test: Pattern = re.compile(r".*-\d+\.(txt|toml)")
+        self.follow_test: Pattern = re.compile(r".*-[2-9]\d*\.(txt|toml)")
```

Comment:
for now I mostly wanted to make the current mechanism more generic. For now this is not supposed to work for more than 2 files

Reply:
I think this could be made more generic with something like:

```suggestion
        self.follow_test: Pattern = re.compile(r".*-([2-9]|[1-9]\d+)\.(txt|toml)")
```

Reply:
Agreed to defer any work on supporting more than 2 test files until later